File: python/day-23/part2.py
from collections import deque, defaultdict

grid = open(0).read().splitlines()

graph = defaultdict(dict)

end = (len(grid) - 1, len(grid[0]) - 2)
queue = deque([(0, 1, 1, 0)])
visited = set()
while queue:
    nr, nc, dr, dc = queue.popleft()
    visited.add((nr, nc, dr, dc))
    pr, pc = nr, nc
    cr, cc = nr + dr, nc + dc

    length = 1
    dirs = [
        (dr, dc) for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)] if
        0 <= cr + dr < len(grid) and
        0 <= cc + dc < len(grid[0]) and
        grid[cr + dr][cc + dc] != "#" and
        (cr + dr, cc + dc) != (pr, pc)
    ]

    while len(dirs) == 1:
        length += 1
        dr, dc = dirs[0]
        pr, pc = cr, cc
        cr, cc = cr + dr, cc + dc

        dirs = [
            (dr, dc) for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)] if
            0 <= cr + dr < len(grid) and
            0 <= cc + dc < len(grid[0]) and
            grid[cr + dr][cc + dc] != "#" and
            (cr + dr, cc + dc) != (pr, pc)
        ]

    if len(dirs) == 0:
        if (cr, cc) == end:
            graph[(nr, nc)][(cr, cc)] = length
            graph[(cr, cc)][(nr, nc)] = length
        continue

    graph[(nr, nc)][(cr, cc)] = length
    graph[(cr, cc)][(nr, nc)] = length
    for dr, dc in dirs:
        if (cr, cc, dr, dc) not in visited:
            queue.append((cr, cc, dr, dc))

graph = dict(graph)

def find_max_cost_path(start, end, visited=None, path_cost=0, max_cost=0, max_path=None):
    if visited is None:
        visited = set()
    if max_path is None:
        max_path = []

    visited.add(start)

    if start == end:
        if path_cost > max_cost:
            max_cost = path_cost
            max_path = list(visited)
    else:
        for neighbor in graph[start]:
            if neighbor in visited:
                continue
            new_cost = path_cost + graph[start][neighbor]

            new_max_cost, new_max_path = find_max_cost_path(neighbor, end, visited.copy(), new_cost, max_cost, max_path)
            if new_max_cost > max_cost:
                max_cost = new_max_cost
                max_path = new_max_path

    return max_cost, max_path

max_cost, max_path = find_max_cost_path((0, 1), end)

print(max_cost, max_path)

File: rust/day-23/src/part2.rs
use std::collections::{HashMap, HashSet, VecDeque};

/// Compute the cost of the maximum cost path from the `start` node to the `end` node
///
/// > **panics** if there is no path from `start` to `end`
fn _find_max_cost(
    graph: &HashMap<[usize; 2], HashMap<[usize; 2], u32>>,
    start: [usize; 2],
    end: [usize; 2],
    mut visited: HashSet<[usize; 2]>,
    path_cost: u32,
    mut max_cost: u32,
) -> u32 {
    visited.insert(start);

    if start == end {
        return max_cost.max(path_cost);
    }

    for (&neighbour, &cost) in graph.get(&start).unwrap() {
        if visited.contains(&neighbour) {
            continue;
        }

        let new_path_cost = path_cost + cost;
        let new_max_cost = _find_max_cost(
            graph,
            neighbour,
            end,
            visited.clone(),
            new_path_cost,
            max_cost,
        );

        max_cost = max_cost.max(new_max_cost);
    }

    return max_cost;
}

fn find_max_cost(
    graph: &HashMap<[usize; 2], HashMap<[usize; 2], u32>>,
    start: [usize; 2],
    end: [usize; 2],
) -> u32 {
    _find_max_cost(graph, start, end, HashSet::new(), 0, 0)
}

/// Compute maze intersection graph from maze grid from `start` to `end`
fn compute_graph(
    grid: &Vec<Vec<char>>,
    start: [usize; 2],
    end: [usize; 2],
) -> HashMap<[usize; 2], HashMap<[usize; 2], u32>> {
    let mut queue: VecDeque<([usize; 2], [isize; 2])> = VecDeque::from_iter([(start, [1, 0])]);
    let mut visited: HashSet<([usize; 2], [isize; 2])> = HashSet::new();
    let mut graph: HashMap<[usize; 2], HashMap<[usize; 2], u32>> = HashMap::new();

    while let Some(([nr, nc], [mut dr, mut dc])) = queue.pop_front() {
        visited.insert(([nr, nc], [dr, dc]));

        let (mut pr, mut pc) = (nr, nc);
        let (Some(mut cr), Some(mut cc)) = (nr.checked_add_signed(dr), nc.checked_add_signed(dc))
        else {
            continue;
        };

        let mut length = 1;

        let mut dirs: Vec<[isize; 2]> = [[0, 1], [0, -1], [-1, 0], [1, 0]]
            .iter()
            .filter(
                |[dr, dc]| match (cr.checked_add_signed(*dr), cc.checked_add_signed(*dc)) {
                    (Some(nr), Some(nc))
                        if nr < grid.len()
                            && nc < grid[0].len()
                            && grid[nr][nc] != '#'
                            && [nr, nc] != [pr, pc] =>
                    {
                        true
                    }
                    _ => false,
                },
            )
            .cloned()
            .collect::<Vec<_>>();

        while dirs.len() == 1 {
            length += 1;
            [dr, dc] = dirs[0];
            [pr, pc] = [cr, cc];
            [cr, cc] = [(cr as isize + dr) as usize, (cc as isize + dc) as usize];

            dirs = [[0, 1], [0, -1], [-1, 0], [1, 0]]
                .iter()
                .filter(
                    |[dr, dc]| match (cr.checked_add_signed(*dr), cc.checked_add_signed(*dc)) {
                        (Some(nr), Some(nc))
                            if nr < grid.len()
                                && nc < grid[0].len()
                                && grid[nr][nc] != '#'
                                && [nr, nc] != [pr, pc] =>
                        {
                            true
                        }
                        _ => false,
                    },
                )
                .cloned()
                .collect::<Vec<_>>();
        }

        if dirs.len() == 0 {
            if [cr, cc] == end {
                graph.entry([nr, nc]).or_default().insert([cr, cc], length);
                graph.entry([cr, cc]).or_default().insert([nr, nc], length);
            }
            continue;
        }

        graph.entry([nr, nc]).or_default().insert([cr, cc], length);
        graph.entry([cr, cc]).or_default().insert([nr, nc], length);

        for [dr, dc] in dirs {
            if !visited.contains(&([cr, cc], [dr, dc])) {
                queue.push_back(([cr, cc], [dr, dc]));
            }
        }
    }

    graph
}

pub fn solve(input: &str) -> u32 {
    let grid: Vec<Vec<char>> = input.lines().map(|row| row.chars().collect()).collect();

    let start: [usize; 2] = [0, 1];
    let end: [usize; 2] = [grid.len() - 1, grid.len() - 2];

    let graph = compute_graph(&grid, start, end);

    find_max_cost(&graph, start, end)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solve() {
        let input = std::fs::read_to_string("inputs/test.txt")
            .expect("ERROR: inputs/test.txt file not found");

        assert_eq!(solve(&input), 154);
    }
}
